test/statement_types.py

```diff
@@ -0,0 +1,451 @@
+            for n in range(self.number_nested_arrays):
+                lhs_string = "std::vector<" + lhs_string + ">"
+                rhs_string = "{" + ",".join([rhs_string] * self.size) + "}"
+
+            return "{lhs_string} {name} = {rhs_string};".format(lhs_string = lhs_string, name = self.name, rhs_string = rhs_string)
+        except AttributeError:
+            scalar = overload_scalar[self.overload]
+            lhs_string = arg_types[self.inner_type].replace("SCALAR", scalar)
+            for n in range(self.number_nested_arrays):
+                lhs_string = "std::vector<" + lhs_string + ">"
+
+            # This is a bit of a nasty way to convert a python iterable type to C++ code
+            # It doesn't respect the self.size variable
+            value_string = repr(self.value).replace('[', '{').replace(']', '}').replace('(', '{').replace(')', '}')
+            return "{lhs_string} {name} = {value};".format(lhs_string = lhs_string, name = self.name, value = value_string)
+
+class FunctionCallAssign(CppStatement):
+    """Represents a function call and optional assignment"""
+    def __init__(self, function_name, name, *args):
+        """
+        Represents a function call and optional assignment of the result to a variable
+        
+        :param function_name: C++ name of function
+        :param name: C++ name of variable, None if result is not saved
+        :param args: Args to pass to function call
+        """
+        self.function_name = function_name
+        self.name = name
+        self.arg_str = ','.join(arg.name for arg in args)
+        self._is_reverse_mode = any(arg.is_reverse_mode() for arg in args) and self.name is not None
+
+    def is_reverse_mode(self):
+        """
+        Return true if any of the inputs are reverse mode and the output is saved
+
+        This isn't always accurate cause the output could be an int or a boolean
+        or something even if the input is reverse mode
+        """
+        return self._is_reverse_mode
+
+    def is_matrix_like(self):
+        raise Exception("is_matrix_like not implemented for FunctionCallAssign")
+
+    def is_varmat(self):
+        raise Exception("is_varmat not implemented for FunctionCallAssign")
+
+    def cpp(self):
+        """Generate C++"""
+        if self.name:
+            return "auto {name} = stan::math::eval({function_name}({arg_str}));".format(name = self.name, function_name = self.function_name, arg_str = self.arg_str)
+        else:
+            return "{function_name}({arg_str});".format(function_name = self.function_name, arg_str = self.arg_str)
+
+class FunctionCall(FunctionCallAssign):
+    """Represents a function call with no return"""
+    def __init__(self, function_name, *args):
+        """
+        Initialize a function call
+        
+        :param function_name: C++ name of function
+        :param args: Args to pass to function call
+        """
+        FunctionCallAssign.__init__(self, function_name, None, *args)
+
```

Comment:
I don't think you need a separate class here.

Reply:
Removed.